sway/commands/sandbox_socket.c

```diff
@@ -0,0 +1,119 @@
+#define _XOPEN_SOURCE 700 // for strdup
+#include <unistd.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <strings.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <wayland-server-core.h>
+#include "sway/commands.h"
+#include "sway/security.h"
+#include "sway/tree/container.h"
+#include "sway/tree/view.h"
+#include "list.h"
+#include "log.h"
+#include "util.h"
+
+struct sandbox_socket {
+	char* path;
+	struct wl_event_source *src;
+	int fd;
+	uint32_t label_id;
+};
+
+static list_t *sandbox_sockets;
+
+static int fd_accept(int srv_fd, uint32_t mask, void *data) {
+	struct sandbox_socket *sock = data;
+	struct sockaddr_un name;
+	socklen_t slen = sizeof(name);
+
+	int cli_fd = accept(srv_fd, (struct sockaddr *)&name, &slen);
+	if (cli_fd < 0)
+		return 0;
+
+	sway_set_cloexec(cli_fd, true);
+	struct wl_client* client = wl_client_create(server.wl_display, cli_fd);
+	if (client) {
+		security_set_client_label_id(client, sock->label_id);
+	} else {
+		close(cli_fd);
+	}
+
+	return 1;
+}
+
+struct cmd_results *cmd_sandbox_socket(int argc, char **argv) {
+	struct cmd_results *error = NULL;
+	if ((error = checkarg(argc, "sandbox_socket", EXPECTED_AT_LEAST, 2))) {
+		return error;
+	}
+
+	if (!sandbox_sockets)
+		sandbox_sockets = create_list();
+
+	char* op = argv[0];
+	char* path = argv[1];
+
+	if (strcmp(op, "create") == 0) {
+		struct sockaddr_un name = {};
+		size_t path_len = strlen(path) + 1;
+		char* label = NULL;
+		int i = 2;
+		while (i < argc) {
+			if (strcmp(argv[i], "--label") == 0) {
+				if (i + 1 >= argc)
+					return cmd_results_new(CMD_INVALID, "--label requires an argument");
+				label = argv[i + 1];
+				i += 2;
+			} else {
+				return cmd_results_new(CMD_INVALID, "Unknown argument to sandbox_socket");
+			}
+		}
```

Comment:
This is a bit unexpected ordering. I think options usually go before positional arguments. Just to consider:

```suggestion
		int i = 1;
		while (i < argc) {
			if (strcmp(argv[i], "--label") == 0) {
				if (i + 1 >= argc)
					return cmd_results_new(CMD_INVALID, "--label requires an argument");
				label = argv[i + 1];
				i += 2;
			} else if (strcmp(argv[i], "--") == 0) {  // after this any argument should be treated as positional
				++i;
				break;
			} else if (strncmp(argv[i], "-", 1) == 0) {
				return cmd_results_new(CMD_INVALID, "Unknown option to sandbox_socket");
			} else {
				break;  // end of options, now only positional arguments
			}
		}
		char *path = argv[i++];
		if (!path) {
				return cmd_results_new(CMD_INVALID, "Missing mandatory argument \"path\" to sandbox_socket");
		} else if (argv[i]) {
				return cmd_results_new(CMD_INVALID, "Unknown argument to sandbox_socket");
		}
```

Reply:
yeah, this is a saner ordering, not sure why I didn't do that in the first place